core/templates/services/audio-player.service.ts

```diff
@@ -16,162 +16,219 @@
-        _play();
-      },
-      pause: function() {
-        _pause();
-      },
-      stop: function() {
-        _stop();
-      },
-      rewind: function(seconds) {
-        _rewind(seconds);
-      },
-      forward: function(seconds) {
-        _forward(seconds);
-      },
-      getCurrentTime: function() {
-        if (_currentTrack) {
-          return Math.round(_currentTrack.currentTime);
-        } else {
-          return 0;
-        }
-      },
-      getAudioDuration: function() {
-        if (_currentTrack && _currentTrack.audio) {
-          return Math.round(_currentTrack.audio.duration);
-        } else {
-          return 0;
-        }
-      },
-      getProgress: function() {
-        if (!_currentTrack) {
-          return 0;
-        }
-        return _currentTrack.progress;
-      },
-      setProgress: function(progress) {
-        if (_currentTrack) {
-          _currentTrack.progress = progress;
-        }
-      },
-      isPlaying: function() {
-        return Boolean(_currentTrack && !_currentTrack.paused);
-      },
-      isTrackLoaded: function() {
-        return Boolean(_currentTrack);
-      },
-      clear: function() {
-        _currentTrack = null;
-        _currentTrackFilename = null;
-      },
-      get onAutoplayAudio() {
-        return _autoplayAudioEventEmitter;
-      }
-    };
-  }
-]);
+        this._currentTrack.on('end', () => {
+          this._currentTrack = null;
+          this._currentTrackFilename = null;
+          this._lastPauseOrSeekPos = null;
+          this.audioTranslationManagerService.clearSecondaryAudioTranslations();
+        });
+      }, (e) => errorCallback(e)
+    );
+  }
+
+  async loadAsync(filename: string): Promise<void> {
+    return new Promise((resolve, reject) => {
+      this._loadAsync(filename, resolve, reject);
+    });
+  }
+
+  play(): void {
+    if (this.isPlaying()) {
+      return;
+    }
+    if (this._currentTrack !== null) {
+      this._currentTrack.seek(this._lastPauseOrSeekPos);
+    }
+    this.ngZone.runOutsideAngular(() => {
+      interval(500).pipe(takeUntil(
+        this._stopIntervalSubject)).subscribe(() => {
+        this.ngZone.run(() => {
+          this._updateViewEventEmitter.emit();
+        });
+      });
+    });
+    this._currentTrack.play();
+  }
+
+  pause(): void {
+    if (!this.isPlaying()) {
+      return;
+    }
+    this._lastPauseOrSeekPos = this.getCurrentTime();
+    this._currentTrack.pause();
+    this._stopIntervalSubject.next();
+  }
+
+  stop(): void {
+    if (!this._currentTrack) {
+      return;
+    }
+    this._lastPauseOrSeekPos = 0;
+    this._currentTrack.stop();
+    this._stopIntervalSubject.next();
+    this._currentTrackFilename = null;
+    this._currentTrack = null;
+  }
+
+  rewind(seconds: number): void {
+    if (!this._currentTrack) {
+      return;
+    }
+    const currentSeconds = (
+      this._currentTrack.seek());
+    if (typeof currentSeconds !== 'number') {
+      return;
```

Comment:
If we have type defined for Howl then do we need this check? (not sure whether `seek` is expected to return number always)

Reply:
![Screenshot 2021-05-11 at 9 42 09 AM](https://user-images.githubusercontent.com/23002114/117757157-29dc8880-b23d-11eb-8a9d-1e69e72cf56a.png)
It returns both number or the Howl instance (depending on whether we are seeking or getting the current time)